src/engine/QueryExecutionTree.cpp

```diff
@@ -399,38 +355,66 @@ nlohmann::json QueryExecutionTree::writeQLeverJsonTable(
 
+// _____________________________________________________________________________
+
+vector<std::optional<pair<size_t, ResultTable::ResultType>>>
+QueryExecutionTree::buildValidIndices(const vector<string>& selectVars,
+                                      const ResultTable& resultTable) const {
+  vector<std::optional<pair<size_t, ResultTable::ResultType>>> validIndices;
+  for (auto var : selectVars) {
+    constexpr std::string_view prefix = "TEXT(";
+    constexpr size_t prefixLength = prefix.length();
+    if (var.starts_with(prefix)) {
+      var = var.substr(prefixLength, var.rfind(')') - prefixLength);
+    }
+    auto it = getVariableColumns().find(var);
+    if (it != getVariableColumns().end()) {
+      validIndices.emplace_back(pair<size_t, ResultTable::ResultType>(
+          it->second, resultTable.getResultType(it->second)));
+    } else {
+      validIndices.emplace_back(std::nullopt);
+    }
+  }
+  return validIndices;
+}
+
 // _____________________________________________________________________________
 template <QueryExecutionTree::ExportSubFormat format>
 ad_utility::stream_generator::stream_generator QueryExecutionTree::writeTable(
-    size_t from, size_t upperBound,
-    const vector<std::optional<pair<size_t, ResultTable::ResultType>>>
-        validIndices,
-    std::shared_ptr<const ResultTable> resultTable) const {
+    const vector<string>& selectVars, size_t limit, size_t offset) const {
   static_assert(format == ExportSubFormat::BINARY ||
                 format == ExportSubFormat::CSV ||
                 format == ExportSubFormat::TSV);
-  if (!resultTable) {
-    resultTable = getResult();
-  }
+
+  // They may trigger computation (but does not have to).
+  shared_ptr<const ResultTable> resultTable = getResult();
+  LOG(DEBUG) << "Resolving strings for finished binary result...\n";
+  auto validIndices = buildValidIndices(selectVars, *resultTable);
+
   const auto& idTable = resultTable->_idTable;
+  size_t upperBound = std::min<size_t>(offset + limit, idTable.size());
+
   // special case : binary export of IdTable
   if constexpr (format == ExportSubFormat::BINARY) {
-    for (size_t i = from; i < upperBound; ++i) {
-      for (size_t j = 0; j < validIndices.size(); ++j) {
-        if (validIndices[j]) {
-          const auto& val = *validIndices[j];
+    for (size_t i = offset; i < upperBound; ++i) {
+      for (const auto& validIndex : validIndices) {
+        if (validIndex) {
           co_yield std::string_view{
-              reinterpret_cast<const char*>(&idTable(i, val.first)),
+              reinterpret_cast<const char*>(&idTable(i, validIndex->first)),
               sizeof(Id)};
         }
       }
     }
     co_return;
   }
 
-  constexpr char sep = format == ExportSubFormat::TSV ? '\t' : ',';
+  static constexpr char sep = format == ExportSubFormat::TSV ? '\t' : ',';
+  constexpr std::string_view sepView{&sep, 1};
+  // Print header line
+  co_yield boost::algorithm::join(selectVars, sepView);
```

Comment:
Let's both have a look into the differences between `boost::algorithm::join` and `absl::StrJoin`.

Reply:
From what I can see they're pretty much the same

Reply:
Ok,
`boost::algorithm::join` has bad performance, as it does exactly what the absl people state as an antiPattern:
`a+= b`;
`a+= c`;
`a+=d`;
`a+=e`;

Wort case : Each element triggers an allocation, in contrast to first getting the total size of all the strings and calling
"reserve" on the result first, at least for types, which support this.
Not that any of this matters here,
But in general we want to use the preferred pattern, not that somedy else starts to think, that this is ok.